### PR TITLE
Refactor violations schema

### PR DESCRIPTION
**Pull request by JoelCourtney**
## Description

This refactors the schema of constraint violations to move data common to all violations of a given constraint into an outer object. Before, the results of a single constraint were represented as `List<Violation>`, where each violation contained metadata about the constraint, and its own list of gaps. Now they are stored in a single `ConstraintResult` object, which contains constraint metadata and gaps at the top level, and a list of `Violation`s, each of which can contain a list of violating intervals and associated activities.

The reason for this is because we want to be able to associate multiple intervals with a single violation, or a single interval with multiple violations, without duplicating common data. This was not an issue before making a rolling window constraint node, which needs multiple intervals per violation.

## Verification
A few AST tests broke and needed fixing.

- [x] I think there is still an e2e test that will fail. I'll fix it if it does, or add one if not. (edit: indeed)

## Documentation
- [ ] There is an example of the violation schema in the docs that needs to be updated.

## Future work
- [ ] This will require an update to the UI to account for the new schema. Also, currently the UI flattens the list of intervals in each violation into separate violations. This will need to be changed. (https://github.com/NASA-AMMOS/aerie-ui/issues/797)

### CODE REVIEW

File: deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql
Comment (Mythicaeda) [inline deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql]: Minor: comment tweak proposed for clarity/accuracy

```suggestion
-- Theoretically we could use SQL to transform the schemas of existing JSON objects, but since the table is just a cache it is safe to clear it
```

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/ConstraintResult.java
Comment (Mythicaeda) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/ConstraintResult.java]: Since these two are always known during initialization, can they be `final`?

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/Violation.java
Comment (Mythicaeda) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/Violation.java]: Are the `violationIntervals` and `activityInstanceIds` Lists correlated at all?

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOfWindows.java
Comment (Mythicaeda) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOfWindows.java:31]: Would `windows.assignGaps(new Windows(false))` work here?

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java
Comment (Mythicaeda) [inline constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java:723]: Why did the interval change from between 4-6 seconds to 1-2 seconds?

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/ConstraintRepository.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/ConstraintRepository.java]: Minor, nonblocking:
```suggestion
  void insertConstraintRuns(final Map<Long, Constraint> constraintMap, final Map<Long, ConstraintResult> constraintResults,
```

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ConstraintRunRecord.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ConstraintRunRecord.java]: Minor, nonblocking:
```suggestion
  ConstraintResult result
```

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintAction.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintAction.java]: Suggestion: 

```java
final ConstraintResult constraintResult = expression.evaluate(preparedResults, environment);
```

File: deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql
Comment (Mythicaeda) [inline deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql]: Minor: since `constraint_run.violations` can now contain an object without any violations it ought to be renamed.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/LocalConstraintService.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/LocalConstraintService.java]: Minor, nonblocking:
```suggestion
  public void createConstraintRuns(final Map<Long, Constraint> constraintMap, final Map<Long, ConstraintResult> constraintResults, final SimulationDatasetId simulationDatasetId) {
```

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/mocks/StubConstraintService.java
Comment (Mythicaeda) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/mocks/StubConstraintService.java]: Minor, nonblocking:
```suggestion
      final Map<Long, ConstraintResult> contraintResults,
```

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintService.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintService.java]: Minor, nonblocking:
```suggestion
  void createConstraintRuns(Map<Long, Constraint> constraintMap, Map<Long, ConstraintResult> constraintResults, SimulationDatasetId simulationDatasetId);
```

File: deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql
Comment (JoelCourtney) [reply deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql]: <sub>do you mean to suggest that programming is easy</sub>

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/Violation.java
Comment (JoelCourtney) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/Violation.java]: they can't be zipped together if that's what you mean. They are unrelated

**Comment by camargo** [comment]
Created the UI ticket: https://github.com/NASA-AMMOS/aerie-ui/issues/797
Make sure we do not merge this until that ticket has an approved PR.

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOfWindows.java
Comment (JoelCourtney) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ViolationsOfWindows.java:31]: `windows.notEqualTo(windows)` sets everything to false except gaps. We need that to isolate the gaps.

File: deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql
Comment (Mythicaeda) [reply deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql]: <sub>oh sorry let me fix that suggestion</sub>
```suggestion
-- Theoretically we could use SQL to transform the schemas of existing JSON objects, but since the table is just a cache it is safe to clear it
-- programming is hard, ok
```

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java
Comment (JoelCourtney) [reply constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java:723]: The interval in the violation supplier doesn't matter, and it looks like I must have changed it. Its just what is in the violation

File: constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java
Comment (Mythicaeda) [inline constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java:714]: What is this constraint? One violation for every A within 2 seconds of a B?

Comment (JoelCourtney) [reply constraints/src/test/java/gov/nasa/jpl/aerie/constraints/tree/ASTTests.java:714]: This is not a real constraint. It just provides a dummy violation for each pair of activities A and B

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/mocks/StubConstraintService.java
Comment (JoelCourtney) [reply merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/mocks/StubConstraintService.java]: file has been deleted by another pr

File: deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql
Comment (JoelCourtney) [reply deployment/hasura/migrations/AerieMerlin/22_constraint_violation_schema/up.sql]: done

**Comment by JoelCourtney** [comment]
# BEHOLD!

<sub>it builds</sub>

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintAction.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintAction.java]: This check should be moved up to L67, given the error message.

Additionally, once it's moved, the is-null checks for `simDatasetId` can be removed.

It may also need to be documented somewhere that constraints now require a plan to have been simulated at least once to pass, as prior it would empty pass or only process constraints on external datasets.

File: deployment/hasura/migrations/AerieMerlin/25_constraint_violation_schema/up.sql
Comment (Mythicaeda) [inline deployment/hasura/migrations/AerieMerlin/25_constraint_violation_schema/up.sql:6]: The comment change on the column needs to be added here and reverted in the down

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java]: Can you add a bindings e2eTest that raises/parses this?

File: e2e-tests/src/tests/constraints.test.ts
Comment (Mythicaeda) [inline e2e-tests/src/tests/constraints.test.ts]: This test ought to be renamed, since it checks that the correct constraint was checked and not the violations.

File: e2e-tests/src/tests/constraints.test.ts
Comment (Mythicaeda) [inline e2e-tests/src/tests/constraints.test.ts:128]: Can an assertion for the length of `constraintResult.violations` being 1 be added in one of these tests? That check was lost during the schema change.

File: e2e-tests/src/tests/constraints.test.ts
Comment (Mythicaeda) [inline e2e-tests/src/tests/constraints.test.ts:142]: Minor: this test needs to either be renamed or to check the value of `constraintRuns[0].results`.

File: e2e-tests/src/tests/external-datasets.test.ts
Comment (Mythicaeda) [inline e2e-tests/src/tests/external-datasets.test.ts]: comment needs to be updated

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java
Comment (Mythicaeda) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java]: Additionally, a `404` status may make more sense than a `500`, since this error is raised in response to missing data (Ie, the Resource in the Constraint does not exist in the Profile Data, a Plan has no Simulation Datasets)

**Review by Mythicaeda — APPROVED** [review]
🎉

**Comment by jeffpamer** [comment]
I'll do a quick retest of my ui branch, and lets try to get that approved too!

**Review by camargo — APPROVED** [review]
LGTM! Thanks

**Pull request merged**
